### PR TITLE
Adding sensor decorator

### LINKED ISSUE #20323: Add a `@task.sensor` TaskFlow decorator
### Description

Implement a taskflow decorator that uses the decorated function as the poke method.

### Use case/motivation

Here is a sketch of the solution that might work:

[sensor_decorator.txt](https://github.com/apache/airflow/files/7721589/sensor_decorator.txt)

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by mingshi-wang**
Added the @task.sensor decorator to convert a Python function to an instance of the BaseSensorOperator. Example usage of the decorator is:

```
@task.sensor(poke_interval=60, timeout=3600, mode="poke")
def f() -> PokeReturnValue:
    # implement the condition 
    condition_met = ...
    operator_return_value = ...
    return PokeReturnValue(is_done=condition_met, xcom_value=operator_return_value)

```

closes: #20323

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/decorators/__init__.pyi
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]: Individual arguments should be spelled out explicitly (instead of using `kwargs`) since this file is providing editor autocompletion, and `kwargs` is mostly useless for autocompletion.

Also, does this make sense?

```python
@sensor  # No arguments at all!
def func():
    return PokeReturnValue(...)
```

If not, `python_callable` should be removed since that argument is there to specifically support this use case.

Comment (mingshi-wang) [reply airflow/decorators/__init__.pyi]: Thanks @uranusjr! I've updated the signature.

Comment (mingshi-wang) [reply airflow/decorators/__init__.pyi]: The only argument useful for the sensor decorator is "python_callable", but please let me know if there are other arguments that need to be explicit.

File: docs/apache-airflow/tutorial_taskflow_api.rst
Comment (potiuk) [inline docs/apache-airflow/tutorial_taskflow_api.rst]: I think you shoudl mention `PokeReturnValue` here as well. This is a new addition in Airflow 2.3.0 (I see you have it in type, but the doc should be updated as well).

Comment (mingshi-wang) [reply docs/apache-airflow/tutorial_taskflow_api.rst]: Sure!

Comment (potiuk) [reply docs/apache-airflow/tutorial_taskflow_api.rst]: so :) ?

Comment (mingshi-wang) [reply docs/apache-airflow/tutorial_taskflow_api.rst]: Hi @potiuk I have updated the doc. Please have a look again.

**Review by potiuk — APPROVED** [review]

File: airflow/decorators/__init__.pyi
Comment (mingshi-wang) [reply airflow/decorators/__init__.pyi]: @uranusjr I need another +1 to this PR. Could you help take a look again? Thanks!

**Review by uranusjr — APPROVED** [review]
Is it still in time for 2.3?

**Comment by potiuk** [comment]
> Is it still in time for 2.3?

I'd say yes - but leave it up to @jedcunningham @ephraimbuddy to decide.

**Comment by potiuk** [comment]
@ashb @jedcunningham @ephraimbuddy -> shall we add it to 2.3 ? It's a pretty requested feature and seems ready (but I leave it up to you to decide :).

**Comment by ephraimbuddy** [comment]
Fine with me cc: @jedcunningham

**Comment by jedcunningham** [comment]
No objections here 👍

File: airflow/example_dags/example_sensor_decorator.py
Comment (jedcunningham) [inline airflow/example_dags/example_sensor_decorator.py]:
```suggestion
# [END tutorial]
```

**Review by jedcunningham — DISMISSED** [review]

**Comment by ashb** [comment]
Someone will need to check how this interacts with task mapping first please

File: airflow/example_dags/example_sensor_decorator.py
Comment (mingshi-wang) [reply airflow/example_dags/example_sensor_decorator.py]: Fixed as suggested.

**Comment by potiuk** [comment]
I think this one is a bit risky to add for 2.3.0 so let's iterate and merge after we branch off.

**Comment by mingshi-wang** [comment]
@potiuk Is it OK to merge this PR?

**Comment by potiuk** [comment]
@mingshi-wang - docs build is failing (that's something that you should fix first) and rebasing the change would also be useful while you do it. 

Also we are waiting for @jedcunningham 's re-review - just re-requested it.

File: docs/apache-airflow/tutorial_taskflow_api.rst
Comment (uranusjr) [inline docs/apache-airflow/tutorial_taskflow_api.rst]: Don’t do this

File: airflow/decorators/sensor.py
Comment (uranusjr) [inline airflow/decorators/sensor.py]: Why does this not use `PythonSensor`?

File: airflow/decorators/__init__.pyi
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]: This is meaningless. Please see above for examples on how to write a useful type annotation.

File: airflow/decorators/sensor.py
Comment (mingshi-wang) [reply airflow/decorators/sensor.py]: Hi @uranusjr Thanks for the comments! The PR was blocked by CI tests failures which I don't think are related to the PR. Do you have any suggestions on this?

File: docs/apache-airflow/tutorial_taskflow_api.rst
Comment (mingshi-wang) [reply docs/apache-airflow/tutorial_taskflow_api.rst]: Can you be specific?

File: airflow/decorators/sensor.py
Comment (mingshi-wang) [reply airflow/decorators/sensor.py]: I don't see much benefit of having another layer of indirection.

File: airflow/decorators/__init__.pyi
Comment (mingshi-wang) [reply airflow/decorators/__init__.pyi]: Can you be more specific?

File: docs/apache-airflow/tutorial_taskflow_api.rst
Comment (uranusjr) [reply docs/apache-airflow/tutorial_taskflow_api.rst]: Why are you linking to GitHub?

File: airflow/decorators/__init__.pyi
Comment (uranusjr) [reply airflow/decorators/__init__.pyi]: https://airflow.apache.org/docs/apache-airflow/stable/howto/create-custom-decorator.html#optional-adding-ide-auto-completion-support

File: docs/apache-airflow/tutorial_taskflow_api.rst
Comment (mingshi-wang) [reply docs/apache-airflow/tutorial_taskflow_api.rst]: I see. This is probably why breeze build-docs fails in the CI.

File: airflow/decorators/__init__.pyi
Comment (mingshi-wang) [reply airflow/decorators/__init__.pyi]: Will fix!

File: airflow/decorators/__init__.pyi
Comment (uranusjr) [reply airflow/decorators/__init__.pyi]: By not using `PythonSensor`, this implementation looses the ability to pass context variables (such as `ti`) to the callable. The ability of passing in `templates_dict` is also lost.

File: airflow/decorators/sensor.py
Comment (uranusjr) [reply airflow/decorators/sensor.py]: By not using `PythonSensor`, this implementation looses the ability to pass context variables (such as `ti`) to the callable. The ability of passing in `templates_dict` is also lost.

**Comment by AndreGodinho7** [comment]
Any updates? Eagerly waiting for this to be merged and released!

**Comment by potiuk** [comment]
Rebased it to see if the Helm cheart failures were carried from another error.

**Comment by mingshi-wang** [comment]
> 



> Rebased it to see if the Helm cheart failures were carried from another error.

Hey @potiuk The Helm local executor still fails. Could you help me understand what could be the cause of the failure? I cannot find any useful message. Thanks!

**Comment by potiuk** [comment]
I think this was accidental failure - simply a problem with lack of space on the machine running it I rebased it and let's see if it happens again. It might also be that we need to optimise some log handling (it looks like logs were filling the disk space. It also could be that (sometimes it happens)  that the cahnge causes some recursive behaviour that simpy fills up all available disk while running (by producing continuous stream of logs) - in this case that will be a bug to be fixed by you.

**Comment by potiuk** [comment]
(BTW. The useful message about lack of disk space  was  in a "cancelled" K8S executor (cancelled because of running too long while trying to work with missing space). The local one was cancelled by K8s failing (we have "fast fail" in this case). Simply GitHub UI shows such "cancelled" jobs less prominently than those that are "failing".

**Comment by potiuk** [comment]
I think our Public runners are not big enough to handle "full tests" with k8s :( . So I removed "full tests needed" label and closed/reopened it to run a smaller set of tests

**Comment by mingshi-wang** [comment]
> I think our Public runners are not big enough to handle "full tests" with k8s :( . So I removed "full tests needed" label and closed/reopened it to run a smaller set of tests

Thanks for your help!

**Comment by mingshi-wang** [comment]
HI @jedcunningham I have addressed your comments. Can you take a look again? Thanks!

**Comment by mingshi-wang** [comment]
Hi @potiuk will you be ok to merge this PR? I see all tests passed.

**Comment by potiuk** [comment]
We are waiting for @jedcunningham to confirm that he is OK with it - can't merge it without it.

**Comment by mingshi-wang** [comment]
HI @jedcunningham can you please take a look again? Thanks!

**Comment by mingshi-wang** [comment]
> We are waiting for @jedcunningham to confirm that he is OK with it - can't merge it without it.

Hi @potiuk, Jed seems unavailable. Is it possible to get another reviewer?

**Comment by mingshi-wang** [comment]
Hi @jedcunningham - do you have a chance to take a look?

**Comment by mingshi-wang** [comment]
Hi @uranusjr @potiuk I think I need some help merging this PR. @jedcunningham is not available to approve this PR. Would you provide some help wrapping up this PR? Thanks!

**Comment by potiuk** [comment]
I dismissed Jed's review - can you please rebase it @mingshi-wang -  sorry for long delay - I've been on holidays :).

**Comment by potiuk** [comment]
Just one more check @uranusjr @jedcunningham ?

File: airflow/decorators/__init__.pyi
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]: Can you list out more arguments and the docstring explicitly (like other functions above)? This is shown in editor autocomplete, and the current format is not useful.

**Comment by potiuk** [comment]
@mingshi-wang - I'd love to merge it - can you please address the last comments of @uranusjr ?

**Comment by mingshi-wang** [comment]
> @mingshi-wang - I'd love to merge it - can you please address the last comments of @uranusjr ?

Sure will do.

**Comment by mingshi-wang** [comment]
> @mingshi-wang - I'd love to merge it - can you please address the last comments of @uranusjr ?
@potiuk @uranusjr I addressed the comments. Could you please take a look again?

**Review by uranusjr — APPROVED** [review]

**Comment by mingshi-wang** [comment]
> @mingshi-wang - I'd love to merge it - can you please address the last comments of @uranusjr ?

Hi @potiuk are you ok to merge the PR?

**Comment by potiuk** [comment]
🎉 🎉 - finally @mingshi-wang !

File: airflow/decorators/sensor.py
Comment (ashb) [inline airflow/decorators/sensor.py:28]: Oh, this doesn't inherit from DecoratedOperator -- was there any reason for it not?

Comment (ashb) [reply airflow/decorators/sensor.py:28]: (Not a problem I guess, since it's all tested etc.)

**Pull request merged**
